docs/proposals/checks_as_wheels.md

```diff
@@ -0,0 +1,288 @@
+happens for the tests.
+
+## Constraints
+
+Any solution must work with the new agent, working for both the old and the new
+one is preferable.
+
+## Recommended Solution
+
+### Package checks as Python Wheels
+
+The solution proposed is to migrate packaging and distribution of our checks
+from system packages built with Omnibus to _universal Python Wheels_ when possible,
+and _binary Wheels_ when we need to ship non-python artifacts.
+
+The changes needed in the agent to support this additional method of checks
+distribution are the following:
+https://github.com/DataDog/dd-agent/compare/massi/whl-poc
+The proof of concept should be extended to rework a bit the loading logic but
+it already works.
+
+The work needed to migrate a pure Python check to wheels is not that much, see
+the ntp check for an example:
+https://github.com/DataDog/integrations-core/tree/massi/whl-poc/ntp
+
+Moving to wheels wouldn’t solve any possible problem we’re facing in building
+and shipping `integrations-core` and `integrations-extra` but overall it might
+work better, let’s see few examples.
+
+#### Building
+
+Building a pure Python check would be as easy as doing
+`python setup.py bdist_wheel`. In the case of checks that ship binary artifacts
+more work would be needed to set up the build scripts, still the build strategy
+would be the same. A locally built check is basically a zip archive, so it could
+be easily moved around and picked up by an agent. In the case of a pure python
+check, the archives would be platform independent, meaning one could build the
+check on a Mac and copy and install the artifact in a Linux box.
+
+#### Versioning
+
+This is not strongly related to the wheel packaging system, but the overall
+philosophy of decoupling the checks from the agent that comes with the wheels
+might help with versioning. Each check would have its own release cycle and at
+each agent release we would pick and include in the package the desired version
+of each check in integrations-core by invoking `pip install -r` on a special
+requirements.txt file that would be part of the agent repository. That file
+would likely contain the most recent version of each check wheel package but
+this wouldn’t be enforced, what it counts is that the requirements file would be
+the unique source of truth stating which checks are shipped with which agent.
+
+#### Dependencies
+
+We would leverage Python integrated tools to keep dependencies sane.
+For example, pip would refuse to downgrade a check if this breaks the
+dependencies for another check. See the Upgrade paragraph for what should happen
+at every agent upgrade.
+
+We could also provide a dummy Agent wheel package integrations would depend on:
+this would enforce compatibility with the core agent at install time (enforced
+by pip) instead of at runtime (enforced by the gent itself after reading check
+metadata). This would also make manifest files unnecessary since we would use
```

Comment:
Since much more tools exist to handle `setup.py` files it would also make scripting `integrations-core` easier than right now.